File: Coding Ninjas/Easy/Find All Anagrams in a String.py
from collections import Counter

def findAnagramsIndices(str, ptr, n, m):
    c = Counter( ptr )
    ans = []
    
    for i in range( n - m + 1):
        if Counter( str[ i : i + m ] ) == c:
            ans.append( i )
            
    return ans



# 2
from collections import Counter

def findAnagramsIndices( str, ptr, n, m ):
    
    c = Counter( ptr )
    d = Counter( str[ : m ] )
    ans = []
    i = 0
    
    if c == d:
        ans.append( i )
        
    for ch in str[ m : ]:
        d[ str[i] ] -= 1
        d[ch] += 1
        
        if d[ str[i] ] < 1:
            del d[ str[i] ]
        
        i += 1
        
        if c == d:
            ans.append( i )
                        
    return ans

File: Coding Ninjas/Moderate/Find All Anagrams in a String.java
import java.util.*;

public class Solution
{

    public static ArrayList< Integer > findAnagramsIndices
    ( String str, int n, String ptr, int m )
    {
        HashMap< Character, Integer > p = new HashMap<>();
        HashMap< Character, Integer > q = new HashMap<>();
        ArrayList< Integer > ans = new ArrayList<>();
        m--;

        for ( char ch : ptr.toCharArray() )
        {
            p.put( ch, p.getOrDefault( ch, 0 ) + 1 );
        }

        for ( int i = 0; i < m; i++ )
        {
            q.put( str.charAt( i ), q.getOrDefault( str.charAt( i ), 0 ) + 1 );
        }

        for ( int i = m, j = 0; i < n; i++, j++ )
        {
            q.put( str.charAt( i ), q.getOrDefault( str.charAt( i ), 0 ) + 1 ) ;

            if ( p.equals( q ) )
            {
                ans.add( j );
            }

            if ( q.get( str.charAt( j ) ) < 2 )
            {
                q.remove( str.charAt( j ) );
            }
            else
            {
                q.put( str.charAt( j ), q.get( str.charAt( j ) ) - 1 );
            }
        }

        return ans;
    }

}
